### PR TITLE
daemon,snap,overlord/snapstate: in the API get the snap icon using state

### PR DESCRIPTION
**Pull request by pedronis**

### CODE REVIEW

**Comment by pedronis** [comment]
retest this please

File: daemon/api.go [hunk 1]
```diff
@@ -831,21 +831,27 @@ func sideloadSnap(c *Command, r *http.Request) Response {
...
 	}
-	defer lock.Unlock()
 
-	installed, err := (&snappy.Overlord{}).Installed()
-	snaps := snappy.FindSnapsByName(name, installed)
-	_, snap := bestSnap(snaps)
-	if err != nil || snap == nil {
+	cur := snapst.Current()
+	if cur == nil {
 		return NotFound("unable to find snap with name %q", name)
```
Comment (niemeyer) [inline daemon/api.go]: cannot

Comment (niemeyer) [inline daemon/api.go]: Sorry, nevermind. This was already there.

File: snap/info.go [hunk 2]
```diff
@@ -226,3 +228,26 @@ func (app *AppInfo) ServiceFile() string {
 func (app *AppInfo) ServiceSocketFile() string {
 	return filepath.Join(dirs.SnapServicesDir, app.SecurityTag()+".socket")
 }
+
+// InfoWithSide returns the snap information for the installed snap with the given name and given side-info.
+func InfoWithSide(name string, si *SideInfo) (*Info, error) {
```
Comment (niemeyer) [inline snap/info.go]: Should we name this `ReadInfo`?  I think you suggested that before, and it seems quite nice in this package.

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -257,28 +254,7 @@ func Deactivate(s *state.State, snap string) (*state.TaskSet, error) {
...
-		return nil, err
-	}
-
-	info.SideInfo = *si
-
-	return info, nil
-}
-
-var retrieveInfo = retrieveInfoImpl
+var retrieveInfo = snap.InfoWithSide
```
Comment (niemeyer) [inline overlord/snapstate/snapstate.go]: var readInfo = snap.ReadInfo

File: overlord/snapstate/export_test.go [hunk 4]
```diff
@@ -54,5 +54,5 @@ func (m *SnapManager) AddForeignTaskHandlers() {
 
 func ChangeRetrieveInfo(retrieve func(name string, si *snap.SideInfo) (*snap.Info, error)) func() {
 	retrieveInfo = retrieve
-	return func() { retrieveInfo = retrieveInfoImpl }
+	return func() { retrieveInfo = snap.InfoWithSide }
```
Comment (niemeyer) [inline overlord/snapstate/export_test.go]: `MockReadInfo`, `readInfo`, `snap.ReadInfo`

Seem like a nice chance to make it all more consistent.

File: daemon/api.go [hunk 5]
```diff
@@ -831,21 +831,27 @@ func sideloadSnap(c *Command, r *http.Request) Response {
...
 	}
 
-	path := filepath.Clean(snap.Icon())
+	info, err := snap.InfoWithSide(name, cur)
+	if err != nil && err != state.ErrNoState {
```
Comment (niemeyer) [inline daemon/api.go]: This cannot fail with ErrNoState.

File: daemon/api.go [hunk 6]
```diff
@@ -831,21 +831,27 @@ func sideloadSnap(c *Command, r *http.Request) Response {
...
+		return InternalError("cannot read metadata: %v", err)
```
Comment (niemeyer) [inline daemon/api.go]: `cannot read snap details: %s`

File: daemon/api.go (hunk 1 above)
Comment (niemeyer) [inline daemon/api.go]: `s/unable to/cannot/`, since we ended up touching this anyway. Thanks!

**Comment by niemeyer** [comment]
LGTM

**Comment by niemeyer** [comment]
retest this please

**Pull request merged**